components/common/src/templating/config.rs

```diff
@@ -588,34 +565,156 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
+fn set_permissions<T: AsRef<Path>>(path: T, _user: &str, _group: &str) -> hcore::error::Result<()> {
     use crate::hcore::util::win_perm;
 
     win_perm::harden_path(path.as_ref())
 }
 
+/// Entry point to recursively walk the configuration directory and subdirectories to
+/// construct the list of template files
+///
+/// `dir` should be a directory that exists.
+fn load_templates<T>(dir: T) -> Result<TemplateRenderer>
+where
+    T: AsRef<Path>,
+{
+    let mut template = TemplateRenderer::new();
+    if dir.as_ref().is_dir() {
+        load_templates_recurse(&dir, &PathBuf::new(), &mut template)?;
+    } else {
+        debug!(
+            "Configuration directory is not a directory: {:?}",
+            dir.as_ref()
+        );
+    }
+    Ok(template)
+}
+
+/// Helper function for `load_templates` to recurse into the directory structure.
+///
+/// `dir` should be a directory
+fn load_templates_recurse<T>(
+    dir: T,
+    context: &PathBuf,
+    mut template: &mut TemplateRenderer,
+) -> Result<()>
+where
+    T: AsRef<Path>,
+{
+    if let Ok(entries) = std::fs::read_dir(dir) {
+        for entry in entries {
+            if let Ok(entry) = entry {
+                // Skip any entries in the template directory which aren't files or directories
+                match entry.file_type() {
+                    Ok(file_type) => {
+                        if file_type.is_file() {
+                            // We're storing the pathname relative to the input config directory
+                            // as the identifier for the template
+                            let name = entry.file_name().to_string_lossy().into_owned();
+                            let relative_path = context.join(&name);
+
+                            // JW TODO: This error needs improvement. TemplateFileError is too generic.
+                            template
+                                .register_template_file(
+                                    &relative_path.to_string_lossy(),
+                                    &entry.path(),
+                                )
+                                .map_err(Error::TemplateFileError)?;
+                        } else if file_type.is_dir() {
+                            load_templates_recurse(
+                                &entry.path(),
+                                &subdir_path(&context, &entry),
+                                &mut template,
+                            )?;
+                        }
+                    }
+                    Err(e) => {
+                        debug!("Failed to get file metadata for {:?} : {}", entry, e);
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Helper function to construct a new PathBuf for each level of subdir we
+/// iterate into.
+fn subdir_path(context: &PathBuf, entry: &std::fs::DirEntry) -> PathBuf {
+    let mut subdir = context.clone();
```

Comment:
perhaps the context should be mutable to avoid the clones

Reply:
@mwrock I need to do the `clone`s as the context is different for each subdirectory I recurse into.  I could rewrite to not use a `PathBuf` and pass e.g. a vec! of components, which would mean I don't have to `clone()` 

My feeling was that  there shouldn't be that many config files and it only happens on svc load so not a recurring hit.

Reply:
oh right i see. yeah this is pretty micro and not worth a refactor IMO

Reply:
Can should be able to pass a `&Path` instead. When you `join` to a `Path`, you'll get a new, owned`PathBuf`. When you join to a `PathBuf`, you get a totally brand-new `PathBuf`, so by using a `&Path` here, you cut the allocations in half.

By taking advantage of that, I think you can get rid of this function altogether. When you're recurring on a directory, you just pass a `Path` reference to your `PathBuf`.

Reply:
+1.  That's what I was looking for. Thanks!